backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/VdsManager.java

```diff
@@ -1250,9 +1250,21 @@ private void recoverCpuPinning(List<VM> vms) {
             if (pinning != null) {
                 Set<Integer> pinnedCpus = CpuPinningHelper.getAllPinnedPCpus(pinning);
                 synchronized (this) {
-                    cpuTopology.stream()
-                            .filter(cpu -> pinnedCpus.contains(cpu.getCpu()))
-                            .forEach(cpu -> cpu.pinVm(vm.getId(), vm.getCpuPinningPolicy()));
+                    if (vm.getCpuPinningPolicy() == CpuPinningPolicy.ISOLATE_THREADS) {
+                        for (int pCpu : pinnedCpus) {
+                            cpuTopology.stream()
+                                    .filter(cpu -> cpu.getCpu() == pCpu)
+                                    .findFirst()
+                                    .ifPresent(cpuUnit -> cpuTopology.stream()
+                                            .filter(cpu -> cpuUnit.getSocket() == cpu.getSocket() &&
+                                                    cpuUnit.getCore() == cpu.getCore())
+                                            .forEach(cpu -> cpu.pinVm(vm.getId(), vm.getCpuPinningPolicy())));
+                        }
+                    } else {
+                        cpuTopology.stream()
+                                .filter(cpu -> pinnedCpus.contains(cpu.getCpu()))
+                                .forEach(cpu -> cpu.pinVm(vm.getId(), vm.getCpuPinningPolicy()));
+                    }
```

Comment:
Could we "merge" those two if branches so that it is more clear what the actual difference is when using the isolate threads? The basic algorithm could be the same for both and maybe easier implemented with traditional iteration than the streams.
1. iterate over pCPUs and find the cpu in the cpu topology
2. for isolate threads, add all siblings
3. pin all found cpus

Reply:
TBH, I think it's much more uglier.

```
                    List<VdsCpuUnit> vdsCpuUnits = new ArrayList<>();
                    for (int pCpu : pinnedCpus) {
                        VdsCpuUnit cpu = null;
                        for (VdsCpuUnit vdsCpuUnit : cpuTopology) {
                            if (vdsCpuUnit.getCpu() == pCpu) {
                                cpu = vdsCpuUnit;
                                break;
                            }
                        }
                        if (cpu == null) {
                            return;
                        }
                        if (vm.getCpuPinningPolicy() == CpuPinningPolicy.ISOLATE_THREADS) {
                            for (VdsCpuUnit vdsCpuUnit : cpuTopology) {
                                if (vdsCpuUnit.getSocket() == cpu.getSocket() &&
                                        vdsCpuUnit.getCore() == cpu.getCore()) {
                                    vdsCpuUnits.add(vdsCpuUnit);
                                }
                            }
                        } else {
                            vdsCpuUnits.add(cpu);
                        }
                    }
                    vdsCpuUnits.forEach(cpu -> cpu.pinVm(vm.getId(), vm.getCpuPinningPolicy()));
```

Reply:
I agree that the `for` is uglier in this case than the streams. 
As for the merging those two if branches into one I think it is still worth, especially if you extracted
```
for (VdsCpuUnit vdsCpuUnit : cpuTopology) {
    if (vdsCpuUnit.getSocket() == cpu.getSocket() &&
        vdsCpuUnit.getCore() == cpu.getCore()) {
        vdsCpuUnits.add(vdsCpuUnit);
    }
}
```
to a separate method called addWithThreadSiblings. Then the if would be like:
```
if (vm.getCpuPinningPolicy() == CpuPinningPolicy.ISOLATE_THREADS) {
    addWithThreadSiblings(vdsCpuUnits, cpu)
} else {
    vdsCpuUnits.add(cpu);
}
```